### PR TITLE
[3.x] Add option `modules_enabled_by_default` and minimal CI build

### PR DESCRIPTION
**Pull request by aaronfranke**
3.x version of #51048. This PR creates a new CI build that does a minimal export template build with everything disabled.

### CODE REVIEW

**Comment by akien-mga** [comment]
I think it would be better to backport #41091 first, specifically all modules manually is too much hassle to maintain.

**Review by Scoppio — APPROVED** [review]
Best part of this PR is that it checks itself in the CI.

**Comment by akien-mga** [comment]
> I think it would be better to backport #41091 first, specifically all modules manually is too much hassle to maintain.

I still think this should be done. The current hardcoded list is error prone (and indeed it includes a number of modules which no longer exist or have been renamed, and lack some of those renamed modules), and maintaining it is more hassle than this additional CI build brings in value IMO.

**Comment by aaronfranke** [comment]
Updated the PR to include backporting `modules_enabled_by_default`. This was easier than expected, there were only a few conflicts. I also set @totlmstr as the author since most of the code is his.

File: SConstruct [hunk 1]
```diff
@@ -302,16 +303,21 @@ for path in module_search_paths:
...
+        import config
+
+        try:
+            enabled = config.is_enabled()
+        except AttributeError:
+            pass
+        sys.path.remove(path)
+        sys.modules.pop("config")
+    else:
+        enabled = False
```
Comment (AThousandShips) [inline SConstruct:319]:
```suggestion
    sys.path.insert(0, path)
    import config

    if env_base["modules_enabled_by_default"]:
        enabled = True
        try:
            enabled = config.is_enabled()
        except AttributeError:
            pass
    else:
        enabled = False

    sys.path.remove(path)
    sys.modules.pop("config")
```
To match changes done in 4.x

File: .github/workflows/linux_builds.yml [hunk 2]
```diff
@@ -48,6 +48,16 @@ jobs:
             build-mono: false
             artifact: true
 
+          - name: Minimal Template (target=release, tools=no, everything disabled)
+            cache-name: linux-template-minimal
+            target: release
+            tools: false
+            tests: false
+            sconsflags: >
+              disable_3d=yes disable_advanced_gui=yes deprecated=no minizip=no xaudio2=no
```
Comment (AThousandShips) [inline .github/workflows/linux_builds.yml:57]:
```suggestion
              disable_3d=yes disable_advanced_gui=yes deprecated=no minizip=no
```
Unnecessary as it's already disabled by default

File: .github/workflows/linux_builds.yml [hunk 3]
```diff
@@ -48,6 +48,16 @@ jobs:
...
+              modules_enabled_by_default=no
+            artifact: true
```
Comment (AThousandShips) [inline .github/workflows/linux_builds.yml:59]:
```suggestion
            build-mono: false
            artifact: true
```
Needed as well AFAIK

**Review by AThousandShips — APPROVED** [review]
Will need some fixes to run properly, but very much needed and tested my own version before I realized this PR existed and it works well (assuming the build failures are fixed, see below)

See, for fixes needed to pass:
* https://github.com/godotengine/godot/pull/89565

File: .github/workflows/linux_builds.yml (hunk 3 above)
Comment (AThousandShips) [inline .github/workflows/linux_builds.yml:59]: This is the configuration I used which worked well and matches the one above
```suggestion
          - name: Minimal template (target=release, tools=no, everything disabled)
            cache-name: linux-template-minimal
            target: release
            tools: false
            sconsflags: modules_enabled_by_default=no disable_3d=yes disable_advanced_gui=yes deprecated=no minizip=no debug_symbols=no
            build-mono: false
            artifact: true
```

**Comment by AThousandShips** [comment]
You can rebase now the build should be functioning now

File: .github/workflows/linux_builds.yml [hunk 4]
```diff
@@ -48,6 +48,14 @@ jobs:
             build-mono: false
             artifact: true
 
+          - name: Minimal template (target=release, tools=no, everything disabled)
+            cache-name: linux-template-minimal
+            target: release
+            tools: false
+            sconsflags: modules_enabled_by_default=no disable_3d=yes disable_advanced_gui=yes deprecated=no minizip=no debug_symbols=no
+            build-mono: false
+            artifact: true
```
[the remark this answers is not in the document]
Comment (aaronfranke) [reply .github/workflows/linux_builds.yml]: Good catch. The artifact is useless, I'll remove it.

**Review by akien-mga — APPROVED** [review]

**Pull request merged**